packages/nitro-protocol/contracts/examples/HashLock.sol

```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.7.4;
+pragma experimental ABIEncoderV2;
+
+import '../interfaces/IForceMoveApp.sol';
+import '../Outcome.sol';
+
+/**
+ * @dev The HashLock contract complies with the ForceMoveApp interface and implements a hashlocked payment
+ */
+contract HashLock is IForceMoveApp {
+    struct AppData {
+        bytes32 h;
+        bytes preImage;
+    }
+
+    function validTransition(
+        VariablePart memory a,
+        VariablePart memory b,
+        uint48 turnNumB,
+        uint256
+    ) public override pure returns (bool) {
+        // is this the first and only swap?
+        require(turnNumB == 4, 'turnNumB != 4');
+
+        // Decode variables.
+        // Assumptions:
+        //  - single asset in this channel
+        //  - two parties in this channel
+        //  - not a "guarantee" channel (c.f. Nitro paper)
+        Outcome.AllocationItem[] memory allocationA = decode2PartyAllocation(a.outcome);
+        Outcome.AllocationItem[] memory allocationB = decode2PartyAllocation(b.outcome);
+        bytes memory preImage = abi.decode(b.appData, (AppData)).preImage;
+        bytes32 h = abi.decode(a.appData, (AppData)).h;
+
+        // is the preimage correct?
+        require(keccak256(preImage) == h, 'Incorrect preimage');
```

Comment:
We could use `sha256` hash function here, instead of `keccak256`. This would give better compatibility with chains such as bitcoin.

Reply:
1dd945b

Reply:
What's the gas cost of `sha256` compared to `keccak256`?

Reply:
It is significantly higher https://ethereum.stackexchange.com/questions/3184/what-is-the-cheapest-hash-function-available-in-solidity.

Reply:
IMO this is the kind of thing that should be recorded in either the code or the git history ([reference](https://mtlynch.io/code-review-love/#4-answer-questions-with-the-code-itself)).

Neither `sha256` nor `keccak256` is an obvious choice, and it's worthwhile indicating why one was chosen over the other.

As an 💡 : the contract could allow the application data to specify which hash function to use. This makes both the appData and the transition logic complex, at the benefit of specializing and saving gas in EVM chains, or generalizing at the expense of higher gas cost in more generic chains.

Reply:
Good point, I'll add some explanatory comments in the source code before merging this.

It's a good idea bout declaring the has function in the app data. Another one is to just to have two contracts deployed (one using each hash function) which would save on the branching logic inside the contract. I think we could tackle this in future work.

Reply:
Yep, I think it's best to commit to one hash right now!

Reply:
957b86d